### PR TITLE
fix: pass AWS_REGION to etcd-manager

### LINKED ISSUE #6098: protokube not specify the AWS_REGION in the manifest of etcd-manager
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**
Version 1.10.0

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
1.9.3

**3. What cloud provider are you using?**
AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**
Follow the instructions in https://github.com/kubernetes/kops/blob/master/docs/etcd/manager.md.

**5. What happened after the commands executed?**
`etcd-manager` failed to start because of wrong default region `us-east-1`.

**6. What did you expect to happen?**
`etcd-manager` should locate the bucket in `cn-north-1` region.

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**
```
apiVersion: kops/v1alpha2
kind: Cluster
metadata:
  name: kops-test.k8s.local
spec:
  channel: stable
  cloudProvider: aws
  configBase: s3://kops-state-store/kops-test.k8s.local
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master-cn-north-1b-1
      name: "1"
    - instanceGroup: master-cn-north-1b-2
      name: "2"
    - instanceGroup: master-cn-north-1b-3
      name: "3"
    manager:
      image: kopeio/etcd-manager:latest
    name: main
  - etcdMembers:
    - instanceGroup: master-cn-north-1b-1
      name: "1"
    - instanceGroup: master-cn-north-1b-2
      name: "2"
    - instanceGroup: master-cn-north-1b-3
      name: "3"
    manager:
      image: kopeio/etcd-manager:latest
    name: events
  kubernetesVersion: https://s3.cn-north-1.amazonaws.com.cn/kops-asset/kubernetes/release/v1.9.3
  masterInternalName: api.internal.kops-test.k8s.local
  masterPublicName: api.kops-test.k8s.local
  topology:
    dns:
      type: Public
    masters: public
    nodes: public
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**
In fact, the bug is unrelated to `kops`, but related to `protokube`, so I guess I should paste the logs of `etcd-manager` here:
```
I1119 09:43:22.730080       1 s3context.go:154] unable to get bucket location from region "us-east-1"; scanning all regions: InvalidToken: The provided token is malformed or otherwise invalid.
        status code: 400, request id: <redacted>, host id: <redacted>
F1119 09:43:24.591170       1 main.go:304] error performing scan for legacy data: error marking cluster as created: error creating cluster-creation marker file s3://kops-state-store/kops-test
.k8s.local/backups/etcd/main/control/etcd-cluster-created: Unable to list AWS regions: AuthFailure: AWS was not able to validate the provided access credentials
        status code: 401, request id: <redacted>, host id: <redacted>
```
**9. Anything else do we need to know?**

### PR DESCRIPTION
**Pull request by knight42**
Fixes #6098

### CODE REVIEW

**Comment by rdrgmnzs** [comment]
/remove-lgtm

We actually probably need a better way to pass the region to etcd-manager. The region the s3 bucket is in may not always be the region the cluster is running in.

@justinsb what do you think about setting this so that is defaults to the region the bucket is in but so that the user can overwrite the setting through the spec?

**Comment by knight42** [comment]
ping @justinsb 

Any suggestion?

File: pkg/model/components/etcdmanager/model.go [hunk 1]
```diff
@@ -342,6 +342,9 @@ func (b *EtcdManagerBuilder) buildPod(etcdCluster *kops.EtcdClusterSpec) (*v1.Po
 				awsup.TagNameRolePrefix + "master=1",
 			}
 			config.VolumeNameTag = awsup.TagNameEtcdClusterPrefix + etcdCluster.Name
+			container.Env = []v1.EnvVar{
+				{Name: "AWS_REGION", Value: os.Getenv("AWS_REGION")},
```
Comment (justinsb) [inline pkg/model/components/etcdmanager/model.go]: Is AWS_REGION set to the state store bucket?  I thought we decided it wasn't necessarily the case.

I think we need to explicitly figure out the region of the bucket, and pass that in.

**Comment by justinsb** [comment]
Thanks for the fix - I think it's the right idea.  What we need to be sure of though is that the AWS_REGION we pass in is indeed the correct region, and I don't think that is necessarily the AWS_REGION that is set for kops.  (I think for example you can just set the region in the aws config file, and never set AWS_REGION).

We also shouldn't really set AWS_REGION if AWS_REGION isn't set locally, and in particular not when we're on GCE :-)

File: pkg/model/components/etcdmanager/model.go (hunk 1 above)
Comment (knight42) [reply pkg/model/components/etcdmanager/model.go]: @justinsb If we could infer the region of the bucket that would be best, but the path to the state store bucket is often written in `S3Uri`, like `s3://kops-state-store`, one may not figure out the region easily.

**Comment by knight42** [comment]
> What we need to be sure of though is that the AWS_REGION we pass in is indeed the correct region, and I don't think that is necessarily the AWS_REGION that is set for kops. 

As mentioned above, the path to the state store bucket is often written in `S3Uri`, which means the bucket is usually in the same region of the EC2 instances, which means in most of the cases, AWS_REGION is indeed the correct region.

> We also shouldn't really set AWS_REGION if AWS_REGION isn't set locally, and in particular not when we're on GCE :-)

Given the [context](https://github.com/kubernetes/kops/pull/6099/files#diff-cc02d582f0c8a1fc88c16dedef2e30edR336), we could see that only if the cloud provider was AWS, we would set the AWS_REGION.

**Comment by knight42** [comment]
@rdrgmnzs @justinsb any updates on this?

**Comment by bksteiny** [comment]
I am also running into this problem, and this code fixed my issue while testing the 1.12 beta. If I understand the problem correctly, the current Kops implementation of etcd-manager introduces a breaking change and can't be used if the s3 bucket for a cluster is either in AWS China or (assuming) AWS Gov, right? I can upgrade/revert my test cluster all day with Kops 1.11.x in AWS China, but not with the current implementation in Kops 1.12.

@justinsb - instead of defaulting to the region the cluster is in, are you suggesting the bucket region be explicitly defined in either the cluster spec or some other environment variable? Are you open to @rdrgmnzs's suggestion by defaulting to the same region as the cluster unless otherwise specified somewhere (assuming the cluster spec)?

It sounds like folks are open to resolving this because it is definitely a blocker, but we need to know the preferred way to address it.

**Comment by mikesplain** [comment]
Just a quick update for watchers, we did discuss this again today in office hours.  Currently we're going to hold the release for this and I believe @justinsb will chime in shortly.  If this requires a bunch of rework, we're going to plan to cut 1.12 and add this in a patch or 1.13, whichever comes first.

**Comment by bksteiny** [comment]
Thanks for the update, @mikesplain!

**Comment by knight42** [comment]
Close in favor of #6943

**Pull request closed without merging**